### PR TITLE
ARROW-1415: [GLib] Support date32 and date64

### PR DESCRIPTION
**Pull request by kou**

### CODE REVIEW

File: c_glib/arrow-glib/array-builder.cpp [hunk 1]
```diff
@@ -1158,6 +1164,150 @@ garrow_string_array_builder_append(GArrowStringArrayBuilder *builder,
...
+  auto builder = garrow_array_builder_new(arrow::date64(),
+                                          NULL,
+                                          "[date64-array-builder][new]");
+  return GARROW_DATE64_ARRAY_BUILDER(builder);
+}
+
+/**
+ * garrow_date64_array_builder_append:
+ * @builder: A #GArrowDate64ArrayBuilder.
+ * @value: The number of days since UNIX epoch in signed 64bit integer.
```
Comment (wesm) [inline c_glib/arrow-glib/array-builder.cpp]: Milliseconds since epoch

File: c_glib/arrow-glib/array-builder.h [hunk 2]
```diff
@@ -782,6 +782,108 @@ gboolean garrow_string_array_builder_append(GArrowStringArrayBuilder *builder,
...
+#define GARROW_IS_DATE32_ARRAY_BUILDER(obj)                             \
+  (G_TYPE_CHECK_INSTANCE_TYPE((obj),                                    \
+                              GARROW_TYPE_DATE32_ARRAY_BUILDER))
+#define GARROW_IS_DATE32_ARRAY_BUILDER_CLASS(klass)             \
+  (G_TYPE_CHECK_CLASS_TYPE((klass),                             \
+                           GARROW_TYPE_DATE32_ARRAY_BUILDER))
+#define GARROW_DATE32_ARRAY_BUILDER_GET_CLASS(obj)              \
+  (G_TYPE_INSTANCE_GET_CLASS((obj),                             \
+                             GARROW_TYPE_DATE32_ARRAY_BUILDER,  \
+                             GArrowDate32ArrayBuilderClass))
```
Comment (wesm) [inline c_glib/arrow-glib/array-builder.h]: I wonder if having a code generation script would assist with automating the generation of these boilerplate macros. What do you think?

File: c_glib/arrow-glib/data-type.h [hunk 3]
```diff
@@ -692,6 +692,94 @@ GType                 garrow_string_data_type_get_type (void) G_GNUC_CONST;
...
+#define GARROW_IS_DATE64_DATA_TYPE(obj)                         \
+  (G_TYPE_CHECK_INSTANCE_TYPE((obj),                            \
+                              GARROW_TYPE_DATE64_DATA_TYPE))
+#define GARROW_IS_DATE64_DATA_TYPE_CLASS(klass)                 \
+  (G_TYPE_CHECK_CLASS_TYPE((klass),                             \
+                           GARROW_TYPE_DATE64_DATA_TYPE))
+#define GARROW_DATE64_DATA_TYPE_GET_CLASS(obj)                  \
+  (G_TYPE_INSTANCE_GET_CLASS((obj),                             \
+                             GARROW_TYPE_DATE64_DATA_TYPE,      \
+                             GArrowDate64DataTypeClass))
```
Comment (wesm) [inline c_glib/arrow-glib/data-type.h]: Same questions here, and in array.h, about automating boilerplate generation

**Review by wesm — APPROVED** [review]
Minor typo and comment, but otherwise LGTM

**Comment by wesm** [comment]
Thanks @kou, merging this

File: c_glib/arrow-glib/array-builder.cpp (hunk 1 above)
Comment (kou) [reply c_glib/arrow-glib/array-builder.cpp]: Oops!

**Comment by kou** [comment]
Thanks! Too fast!

File: c_glib/arrow-glib/array-builder.h (hunk 2 above)
Comment (kou) [reply c_glib/arrow-glib/array-builder.h]: It's reasonable.
I'll talk about with GLib developers. If they also think it's useful, we should share it.

Comment (kou) [reply c_glib/arrow-glib/array-builder.h]: I heard at gtk-list mailing list and a developer tells me a convenient API: https://mail.gnome.org/archives/gtk-list/2017-September/msg00031.html

I'll migrate to the API later.

**Pull request closed without merging**